docs/dev/relay_add_op.rst

```diff
@@ -0,0 +1,148 @@
+Adding an Operator to Relay
+===========================
+
+In order to use TVM operators from within the Relay IR, the
+operators need to be registered in Relay in order to ensure
+that they will be integrated into Relay's type system.
+
+Registering an operator requires three steps:
+- Using the ``RELAY_REGISTER_OP`` macro in C++ to
+register the operator's arity and type information
+- Defining a C++ function to produce a call node for the
+operator and registering a Python API hook for the function
+- Wrapping the above Python API hook in a neater interface
+
+The file ``src/relay/op/tensor/elemwise.cc`` provides
+examples of the first two steps, while
+``python/tvm/relay/op/tensor.py`` gives examples of the
+last.
+
+Registering an Operator
+-----------------------
+
+TVM already has an operator registry, but Relay cannot properly
+incorporate TVM operators without additional type information.
+
+To allow for flexibility in registering operators and greater
+expressivity and granularity in expressing types in Relay, operators
+are typed using relations between input and output types. These relations
+are represented as functions that take in a list of input types and
+output types (any of these types may be incomplete) and return a list
+of input and output types that satisfies the relation. Essentially, a
+relation for an operator can enforce all the necessary typing rules
+(namely by inspecting the input types) in addition to computing the
+output type.
+
+For example, see ``src/relay/op/type_relations.h`` and their
+implementations. E.g., ``BroadcastRel`` takes two input types and an
+output type, checks that they are all tensor types with the same underlyin
+data type, and finally ensures that the shape of the output type is the
+broadcast of the input types' shapes.
+
+It may be necessary to add another type relation to ``type_relations.h``
+if the existing ones do not capture the behavior of the desired operator.
+
+The ``RELAY_REGISTER_OP`` macro in C++ allows a developer
+to specify the following information about an operator in Relay:
+- Arity (number of arguments)
+- Names and descriptions for positional arguments
+- Support level (1 indicating an internal intrinsic, higher numbers
+indicating operators that are not as integral to the framework or are
+supported externally)
+- A type relation for the operator
+
+The below example is from ``elemwise.cc`` and uses a broadcasting
+add for tensors:
+
+.. code:: python

```

Comment:
.. code:: C++?

Reply:
I don't know if RST supports C++, but I can change it to C